File: 220.contains-duplicate-iii.py
#
# @lc app=leetcode id=220 lang=python3
#
# [220] Contains Duplicate III
#

# Time:  O(n * t)
# Space: O(max(k, t))

from typing import List
from collections import OrderedDict


# @lc code=start
class Solution:
    def containsNearbyAlmostDuplicate(self, nums: List[int], k: int, t: int) -> bool:
        window = OrderedDict()
        for num in nums:
            if len(window) > k:
                window.popitem(False)

            bucket = num if not t else num // t
            for n in (
                window.get(bucket - 1),
                window.get(bucket),
                window.get(bucket + 1),
            ):
                if n is not None and abs(num - n) <= t:
                    return True
            window[bucket] = num
        return False


# @lc code=end

File: 220.contains-duplicate-iii.js
/*
 * @lc app=leetcode id=220 lang=javascript
 *
 * [220] Contains Duplicate III
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} k
 * @param {number} t
 * @return {boolean}
 */
var containsNearbyAlmostDuplicate = function (nums, k, t) {
  const window = new Map();
  for (const num of nums) {
    if (window.size > k) {
      window.delete(Array.from(window.keys())[0]);
    }
    const bucket = t ? Math.trunc(num / t) : num;
    for (let i = -1; i < 2; ++i) {
      const n = window.get(bucket + i);
      if (n !== undefined && Math.abs(num - n) <= t) {
        return true;
      }
    }
    window.set(bucket, num);
  }
  return false;
};
// @lc code=end
